File: ontrack-extension-general/src/main/java/net/nemerosa/ontrack/extension/general/LinkPropertyType.java
package net.nemerosa.ontrack.extension.general;

import com.fasterxml.jackson.databind.JsonNode;
import net.nemerosa.ontrack.extension.support.AbstractPropertyType;
import net.nemerosa.ontrack.model.form.Form;
import net.nemerosa.ontrack.model.form.NamedEntries;
import net.nemerosa.ontrack.model.security.ProjectConfig;
import net.nemerosa.ontrack.model.security.SecurityService;
import net.nemerosa.ontrack.model.structure.ProjectEntity;
import net.nemerosa.ontrack.model.structure.ProjectEntityType;
import net.nemerosa.ontrack.model.support.NameValue;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.EnumSet;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

@Component
public class LinkPropertyType extends AbstractPropertyType<LinkProperty> {

    @Autowired
    public LinkPropertyType(GeneralExtensionFeature extensionFeature) {
        super(extensionFeature);
    }

    @Override
    public String getName() {
        return "Links";
    }

    @Override
    public String getDescription() {
        return "List of links.";
    }

    @Override
    public Set<ProjectEntityType> getSupportedEntityTypes() {
        return EnumSet.allOf(ProjectEntityType.class);
    }

    @Override
    public boolean canEdit(ProjectEntity entity, SecurityService securityService) {
        return securityService.isProjectFunctionGranted(entity, ProjectConfig.class);
    }

    @Override
    public boolean canView(ProjectEntity entity, SecurityService securityService) {
        return true;
    }

    @Override
    public Form getEditionForm(ProjectEntity entity, LinkProperty value) {
        return Form.create()
                .with(
                        NamedEntries.of("links")
                                .label("List of links")
                                .nameLabel("Name")
                                .valueLabel("Link")
                                .nameOptional()
                                .addText("Add a link")
                                .help("List of links associated with a name.")
                                .value(value != null ? value.getLinks() : Collections.emptyList())
                )
                ;
    }

    @Override
    public LinkProperty fromClient(JsonNode node) {
        return fromStorage(node);
    }

    @Override
    public LinkProperty fromStorage(JsonNode node) {
        return parse(node, LinkProperty.class);
    }

    @Override
    public String getSearchKey(LinkProperty value) {
        return value.getLinks().stream()
                .map(namedLink -> {
                    if (StringUtils.isNotBlank(namedLink.getName())) {
                        return namedLink.getName();
                    } else {
                        return namedLink.getValue();
                    }
                })
                .collect(Collectors.joining(","));
    }

    @Override
    public LinkProperty replaceValue(LinkProperty value, Function<String, String> replacementFunction) {
        return new LinkProperty(
                value.getLinks().stream()
                        .map(nv -> new NameValue(nv.getName(), replacementFunction.apply(nv.getValue())))
                        .collect(Collectors.toList())
        );
    }
}

File: ontrack-extension-general/src/main/java/net/nemerosa/ontrack/extension/general/LinkPropertyType.kt
package net.nemerosa.ontrack.extension.general

import com.fasterxml.jackson.databind.JsonNode
import net.nemerosa.ontrack.extension.support.AbstractPropertyType
import net.nemerosa.ontrack.model.json.schema.JsonType
import net.nemerosa.ontrack.model.json.schema.JsonTypeBuilder
import net.nemerosa.ontrack.model.json.schema.toType
import net.nemerosa.ontrack.model.security.ProjectConfig
import net.nemerosa.ontrack.model.security.SecurityService
import net.nemerosa.ontrack.model.structure.ProjectEntity
import net.nemerosa.ontrack.model.structure.ProjectEntityType
import net.nemerosa.ontrack.model.structure.PropertySearchArguments
import net.nemerosa.ontrack.model.support.NameValue
import org.apache.commons.lang3.StringUtils
import org.springframework.stereotype.Component
import java.util.*

@Component
class LinkPropertyType(
        extensionFeature: GeneralExtensionFeature
) : AbstractPropertyType<LinkProperty>(extensionFeature) {

    override val name: String = "Links"

    override val description: String = "List of links."

    override val supportedEntityTypes: Set<ProjectEntityType> =
            EnumSet.allOf(ProjectEntityType::class.java)

    override fun createConfigJsonType(jsonTypeBuilder: JsonTypeBuilder): JsonType =
        jsonTypeBuilder.toType(LinkProperty::class)

    override fun canEdit(entity: ProjectEntity, securityService: SecurityService): Boolean =
            securityService.isProjectFunctionGranted(entity, ProjectConfig::class.java)

    override fun canView(entity: ProjectEntity, securityService: SecurityService): Boolean = true

    override fun fromClient(node: JsonNode): LinkProperty {
        return fromStorage(node)
    }

    override fun fromStorage(node: JsonNode): LinkProperty {
        return parse(node, LinkProperty::class)
    }

    override fun replaceValue(value: LinkProperty, replacementFunction: (String) -> String): LinkProperty {
        return LinkProperty(
                value.links.map { nv ->
                    NameValue(nv.name, replacementFunction(nv.value))
                }
        )
    }

    override fun getSearchArguments(token: String): PropertySearchArguments? {
        return PropertySearchArguments(
                jsonContext = "jsonb_array_elements(pp.json->'links') as link",
                jsonCriteria = "link->>'value' ilike :value",
                criteriaParams = mapOf(
                        "value" to "%$token%"
                )
        )
    }

    override fun containsValue(value: LinkProperty, propertyValue: String): Boolean =
            value.links.map {
                it.value
            }.any {
                StringUtils.containsIgnoreCase(it, propertyValue)
            }
}
